x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -249,8 +251,52 @@ export class ContentStream extends Duplex {
     });
   }
 
-  private async flush(size = this.buffer.byteLength) {
-    const chunk = this.buffer.slice(0, size);
+  private async flush(size = this.bytesBuffered) {
+    const buffersToFlush: Buffer[] = [];
+    let partiallyFlushedBuffer: undefined | Buffer;
+    let bytesToFlush = 0;
+
+    /*
+     Loop over each buffer, keeping track of how many bytes we have added
+     to the array of buffers to be flushed. The array of buffers to be flushed
+     contains buffers by reference, not copies. This avoids putting pressure on
+     the CPU for copying buffers or for gc activity. Please profile performance
+     with a large byte configuration and a large number of records (900k+)
+     before changing this code. Config used at time of writing:
+
+      xpack.reporting:
+        csv.maxSizeBytes: 500000000
+        csv.scroll.size: 1000
+
+     At the moment this can put memory pressure on Kibana. Up to 1,1 GB in a dev
+     build. It is not recommended to have overly large max size bytes but we
+     need this code to be as performant as possible.
+    */
+    for (const buffer of this.buffers) {
+      const remainder = size - bytesToFlush;
+      if (remainder < 0) {
+        break;
+      }
+      const chunkedBuffer = remainder > 0 ? buffer.slice(0, remainder) : buffer;
+
+      buffersToFlush.push(chunkedBuffer);
+      bytesToFlush += chunkedBuffer.byteLength;
+
+      // Did we add a partial buffer? Cache the rest and stop looping.
+      if (buffer.byteLength > remainder) {
+        partiallyFlushedBuffer = buffer.slice(remainder);
+        break;
+      }
+    }
+
+    if (this.buffers.length > 0 && buffersToFlush.length === 0) {
```

Comment:
That's a redundant check. The `http.max_content_length` is always a positive value. Or am I missing something?

Reply:
This is to cover the edge cases where the http.max_content_length is a bogus value. I just thought it was a case not covered in the code. I'm happy to remove it though if we have gaurantees that it will always be > 0.